File: src/main/scala/com/exasol/cloudetl/kafka/deserialization/DeserializationFactory.scala
package com.exasol.cloudetl.kafka.deserialization

import com.exasol.cloudetl.kafka.{KafkaConnectorException, KafkaConsumerProperties}
import com.exasol.errorreporting.ExaError

import org.apache.kafka.common.serialization.Deserializer

object DeserializationFactory {

  final case class RecordDeserializers(
    keyDeserializer: Deserializer[Map[FieldSpecification, Seq[Any]]],
    valueDeserializer: Deserializer[Map[FieldSpecification, Seq[Any]]]
  )

  def getSerializers(
    fieldSpecs: Seq[GlobalFieldSpecification],
    kafkaProperties: KafkaConsumerProperties
  ): RecordDeserializers = {

    val keyFieldSpecs = fieldSpecs.flatMap {
      case keySpec: KeySpecification => Option(keySpec)
      case _                         => None
    }

    val valueFieldSpecs = fieldSpecs.flatMap {
      case valueSpec: ValueSpecification => Option(valueSpec)
      case _                             => None
    }

    val keyDeserializer = Option(keyFieldSpecs)
      .filter(_.nonEmpty)
      .map { keyFields =>
        getDeserialization(kafkaProperties.getRecordKeyFormat())
          .getDeserializer(kafkaProperties, keyFields)
      }
      .getOrElse(IgnoreKeyDeserializer)

    val valueDeserializer = getDeserialization(kafkaProperties.getRecordValueFormat())
      .getDeserializer(kafkaProperties, valueFieldSpecs)

    RecordDeserializers(keyDeserializer, valueDeserializer)
  }

  def getDeserialization(format: String): RecordDeserialization =
    format match {
      case "avro"   => AvroDeserialization
      case "json"   => JsonDeserialization
      case "string" => StringDeserialization
      case _ =>
        throw new KafkaConnectorException(
          ExaError
            .messageBuilder("E-KCE-19")
            .message("The format {{FORMAT}} is not supported.", format)
            .mitigation("Please use one of 'avro', 'json' or 'string' formats.")
            .toString()
        )
    }
}

/**
 * If we don't need the key we just return an empty map.
 */
object IgnoreKeyDeserializer extends Deserializer[Map[FieldSpecification, Seq[Any]]] {
  override def deserialize(topic: String, data: Array[Byte]): Map[FieldSpecification, Seq[Any]] =
    Map.empty[FieldSpecification, Seq[Any]]
}

File: src/main/java/com/exasol/cloudetl/kafka/deserialization/DeserializationFactory.java
package com.exasol.cloudetl.kafka.deserialization;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

import org.apache.kafka.common.serialization.Deserializer;

import com.exasol.cloudetl.kafka.*;
import com.exasol.errorreporting.ExaError;

public final class DeserializationFactory {
    private static final Map<String, RecordDeserialization> SUPPORTED_DESERIALIZATIONS = Map.of(
            "avro", AvroDeserialization::getDeserializer,
            "json", JsonDeserialization::getDeserializer,
            "string", StringDeserialization::getDeserializer);

    private DeserializationFactory() {
    }

    public static RecordDeserializers getSerializers(
            final List<GlobalFieldSpecification> fieldSpecs, final KafkaConsumerProperties kafkaProperties) {
        final List<FieldSpecification> keyFieldSpecs = new ArrayList<>();
        final List<FieldSpecification> valueFieldSpecs = new ArrayList<>();
        for (final GlobalFieldSpecification spec : fieldSpecs) {
            if (spec instanceof KeySpecification) {
                keyFieldSpecs.add((FieldSpecification) spec);
            } else if (spec instanceof ValueSpecification) {
                valueFieldSpecs.add((FieldSpecification) spec);
            }
        }
        final Deserializer<Map<FieldSpecification, List<Object>>> keyDeserializer = keyFieldSpecs.isEmpty()
                ? FieldSpecificationSingletons.ignoreKeyDeserializer()
                : getDeserialization(kafkaProperties.getRecordKeyFormat()).getDeserializer(kafkaProperties, keyFieldSpecs);
        final Deserializer<Map<FieldSpecification, List<Object>>> valueDeserializer =
                getDeserialization(kafkaProperties.getRecordValueFormat()).getDeserializer(kafkaProperties, valueFieldSpecs);
        return new RecordDeserializers(keyDeserializer, valueDeserializer);
    }

    public static RecordDeserializers getSerializers(final scala.collection.immutable.Seq<GlobalFieldSpecification> fieldSpecs,
            final KafkaConsumerProperties kafkaProperties) {
        return getSerializers(ScalaCollections.javaList(fieldSpecs), kafkaProperties);
    }

    public static RecordDeserialization getDeserialization(final String format) {
        final RecordDeserialization deserialization = SUPPORTED_DESERIALIZATIONS.get(format);
        if (deserialization == null) {
            throw new KafkaConnectorException(ExaError.messageBuilder("E-KCE-19")
                    .message("The format {{FORMAT}} is not supported.", format)
                    .mitigation("Please use one of {{SUPPORTED_FORMATS}} formats.",
                            String.join(",", SUPPORTED_DESERIALIZATIONS.keySet()))
                    .toString());
        }
        return deserialization;
    }

    public static final class RecordDeserializers {
        public final Deserializer<Map<FieldSpecification, List<Object>>> keyDeserializer;
        public final Deserializer<Map<FieldSpecification, List<Object>>> valueDeserializer;

        public RecordDeserializers(
                final Deserializer<Map<FieldSpecification, List<Object>>> keyDeserializer,
                final Deserializer<Map<FieldSpecification, List<Object>>> valueDeserializer) {
            this.keyDeserializer = keyDeserializer;
            this.valueDeserializer = valueDeserializer;
        }

        public Deserializer<Map<FieldSpecification, List<Object>>> getKeyDeserializer() {
            return this.keyDeserializer;
        }

        public Deserializer<Map<FieldSpecification, List<Object>>> getValueDeserializer() {
            return this.valueDeserializer;
        }

        @Override
        public boolean equals(final Object other) {
            if (!(other instanceof RecordDeserializers)) {
                return false;
            }
            final RecordDeserializers that = (RecordDeserializers) other;
            return Objects.equals(this.keyDeserializer, that.keyDeserializer)
                    && Objects.equals(this.valueDeserializer, that.valueDeserializer);
        }

        @Override
        public int hashCode() {
            return Objects.hash(this.keyDeserializer, this.valueDeserializer);
        }
    }
}
